File: cpp/base/src/vsdk/toolkit/io/image/PicOutputHandle.cpp
#include "vsdk/toolkit/java/io/FileOutputStream.h"
#include "vsdk/toolkit/java/lang/System.h"
#include "vsdk/toolkit/java/util/Formatter.h"
#include "vsdk/toolkit/io/wrapper/PersistenceElement.h"
#include "vsdk/toolkit/io/image/Dkcolor.h"
#include "vsdk/toolkit/io/image/PicOutputHandle.h"

java::String
PicOutputHandle::formatToString(const char *format, va_list arguments) {
    if ( format == nullptr ) {
        return java::String();
    }

    char localBuffer[256];
    va_list argumentsCopy;
    va_copy(argumentsCopy, arguments);
    const int required = java::Formatter::vformat(localBuffer, static_cast<int>(sizeof(localBuffer)), format, argumentsCopy);
    va_end(argumentsCopy);

    if ( required < 0 ) {
        return java::String();
    }
    if ( required < static_cast<int>(sizeof(localBuffer)) ) {
        return java::String(localBuffer);
    }

    char *dynamicBuffer = new char[required + 1];
    va_copy(argumentsCopy, arguments);
    java::Formatter::vformat(dynamicBuffer, required + 1, format, argumentsCopy);
    va_end(argumentsCopy);

    java::String result(dynamicBuffer);
    delete[] dynamicBuffer;
    return result;
}

void
PicOutputHandle::writeFormatted(java::OutputStream *outputStream, const char *format, ...) {
    if ( outputStream == nullptr || format == nullptr ) {
        return;
    }

    va_list arguments;
    va_start(arguments, format);
    java::String text = formatToString(format, arguments);
    va_end(arguments);

    if ( text.isEmpty() ) {
        return;
    }
    vsdk::PersistenceElement::writeBytes(
        *outputStream,
        reinterpret_cast<const unsigned char *>(text.toCString()),
        text.length());
}

PicOutputHandle::PicOutputHandle(const char *filename, int w, int h) {
    ImageOutputHandle::init("high dynamic range PIC", w, h);

    java::File file(filename);
    if ( !file.canWrite() || file.isDirectory() ) {
        outputStream = nullptr;
        java::System::err.printf("Can't open PIC output");
        return;
    }
    outputStream = new java::FileOutputStream(filename);

    writeHeader();
}

PicOutputHandle::~PicOutputHandle() {
    if ( outputStream != nullptr ) {
        outputStream->close();
        delete outputStream;
    }
    outputStream = nullptr;
}

/**
Writes scanline of high-dynamic range radiance data in RGB format
*/
int
PicOutputHandle::writeRadianceRGB(ColorRgbMutable *rgbRadiance) {
    int result = 0;

    if ( outputStream != nullptr ) {
        result = DkColor::writeScan(reinterpret_cast<DK_COLOR *>(rgbRadiance), width, outputStream);
    }

    if ( result ) {
        return width;
    } else {
        // We don't know how many pixels were actually written
        return 0;
    }
}

void
PicOutputHandle::writeHeader() {
    // Simple RADIANCE header
    writeFormatted(outputStream, "#?RADIANCE\n");
    writeFormatted(outputStream, "#RPK PicOutputHandler (compiled %s)\n", __DATE__);
    writeFormatted(outputStream, "FORMAT=32-bit_rle_rgbe\n");
    writeFormatted(outputStream, "\n");
    writeFormatted(outputStream, "-Y %d +X %d\n", height, width);
}

File: java/base/src/main/java/vsdk/toolkit/io/image/PicOutputHandle.java
package vsdk.toolkit.io.image;

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.io.PersistenceElement;

/**
High dynamic range PIC output handle.

Olaf Appeltants, March 2000
*/
public final class PicOutputHandle extends ImageOutputHandle {
    private OutputStream outputStream;

    private static String formatToString(String format, Object... arguments) {
        if (format == null) {
            return "";
        }
        try {
            return String.format(Locale.US, format, arguments);
        }
        catch (Exception ignored) {
            return "";
        }
    }

    private static void writeFormatted(OutputStream outputStream, String format, Object... arguments) {
        if (outputStream == null || format == null) {
            return;
        }

        String text = formatToString(format, arguments);
        if (text.isEmpty()) {
            return;
        }

        byte[] bytes = text.getBytes(StandardCharsets.US_ASCII);
        PersistenceElement.writeBytes(outputStream, bytes, bytes.length);
    }

    public PicOutputHandle(String filename, int w, int h) {
        init("high dynamic range PIC", w, h);

        File file = new File(filename);
        if (!file.canWrite() || file.isDirectory()) {
            outputStream = null;
            System.err.printf("Can't open PIC output");
            return;
        }

        try {
            outputStream = new FileOutputStream(filename);
        }
        catch (Exception ignored) {
            outputStream = null;
            System.err.printf("Can't open PIC output");
            return;
        }

        writeHeader();
    }

    void closeHandle() {
        if (outputStream != null) {
            try {
                outputStream.close();
            }
            catch (Exception ignored) {
            }
        }
        outputStream = null;
    }

    /**
Writes scanline of high-dynamic range radiance data in RGB format
*/
    @Override
    public int writeRadianceRGB(ColorRgb[] rgbRadiance) {
        int result = 0;

        if (outputStream != null) {
            result = DkColor.writeScan(rgbRadiance, width, outputStream);
        }

        if (result != 0) {
            return width;
        }
        else {
            // We don't know how many pixels were actually written
            return 0;
        }
    }

    private void writeHeader() {
        // Simple RADIANCE header
        writeFormatted(outputStream, "#?RADIANCE\n");
        String compileDate = LocalDate.now().format(DateTimeFormatter.ofPattern("MMM dd yyyy", Locale.US));
        writeFormatted(outputStream, "#RPK PicOutputHandler (compiled %s)\n", compileDate);
        writeFormatted(outputStream, "FORMAT=32-bit_rle_rgbe\n");
        writeFormatted(outputStream, "\n");
        writeFormatted(outputStream, "-Y %d +X %d\n", height, width);
    }
}
